### DISCUSSION TITLE
ETA for v1.21.2

### DISCUSSION THREAD

**Discussion by akshaykarangale** [pymupdf/PyMuPDF#2200] [General]
Hello,
First of all thanks for such an awesome library!

I was wondering if there's any timeline on when v1.21.2 would be released. Eagerly waiting to try out the clipping path details that is to be released in this version.

Thanks!

**Comment by JorjMcKie** [comment]
What is your environment - if Linux Python 3.10 or any Windows, I could give you a pre-version wheel.

**Reply by akshaykarangale** [reply]
I am on Ubuntu with Python 3.7.4 and Python 3.8. If the wheel needs a different version, I could make a new one with the given requirements. I am flexible with the environment

**Reply by JorjMcKie** [reply]
Ok, best try this one. Please be aware that it is still under development for some features outside this topic, especially font support in `TextWriter` and `Story`. Changes in a nutshell:
* New data are returned if specifying `extended=True` in `get_drawings()` / `get_drawings()`.
* In any case, even without "extended", a new key "layer" is present which contains the name of any Optional Content Group - or `None`.
* Every dictionay in the list contains the key `level` which reflects the hierarchy established by Clip or Group paths. What **_level_** means can be seen in the output of `mutool trace ...`.
* Clip and Group paths increase the level number. Every subsequent path with a higher level is under the control of previous groups or clips of lower level, see the following output:
```
1  0 group Rect(0.0, 0.0, 343.432861328125, 206.94017028808594)
2   1 clip Rect(3.9179999828338623, 73.03500366210938, 344.0, 199.10101318359375)
3    2 clip Rect(3.9179999828338623, 73.03500366210938, 344.0, 199.10101318359375)
4     3 clip Rect(14.0, 73.03500366210938, 107.0, 193.0)
5      4 clip Rect(14.17199993133545, 73.03500366210938, 106.93399810791016, 192.76600646972656)
6       5 clip Rect(14.17199993133545, 73.03500366210938, 106.93399810791016, 192.76600646972656)
7        6 f Rect(-20.18000030517578, 4.168000221252441, 146.68301391601562, 220.8280029296875)
8     3 clip Rect(211.0, 73.03500366210938, 330.0, 156.0)
9      4 clip Rect(211.18800354003906, 73.03500366210938, 329.25799560546875, 156.0)
a       5 clip Rect(211.18800354003906, 73.03500366210938, 329.25799560546875, 156.0)
b        6 f Rect(186.23800659179688, 4.578000068664551, 353.10101318359375, 221.23800659179688)
```
* Clips have a "scissor" rectangle which is the area, inside which the following paths are living.
* While level is increasing, scissor rects are contained in the previous scissor
* In the above, line 8 (level 3) continues the clip sequence of line 3

Here are two test scripts to produce debugging output: 
[useful-scripts.zip](https://github.com/pymupdf/PyMuPDF/files/10533659/useful-scripts.zip)

* The same result functionally can be produced by new method `page.get_lineart()`. It returns a list-like object that has additional methods and properties for returning the parent clips of any path, or the total number of fill clips, etc.
Plus any path is not a dictionary, but an object. Example session:
```python
In [1]: import fitz
In [2]: doc=fitz.open("clipping_3.pdf")
In [3]: page=doc[0]
In [4]: lineart = page.get_lineart()
In [5]: type(lineart)
Out[5]: fitz.fitz.Page.Drawpathlist
In [6]: lineart[-1]  # the last path from above list
Out[6]: <fitz.fitz.Page.Drawpath at 0x200ad100050>
In [7]: path = lineart[-1]
In [8]: path.rect
Out[8]: Rect(186.23800659179688, 4.578000068664551, 353.10101318359375, 221.23800659179688)
In [9]: path.type
Out[9]: 'f'
In [10]:
In [10]: path.layer
In [11]: path.__dict__
Out[11]:
{'items': [('re',
   Rect(186.23800659179688, 4.578000068664551, 353.10101318359375, 221.23800659179688),
   -1)],
 'type': 'f',
 'even_odd': False,
 'fill_opacity': 1.0,
 'fill': (0.0, 0.5019609928131104, 0.0),
 'rect': Rect(186.23800659179688, 4.578000068664551, 353.10101318359375, 221.23800659179688),
 'seqno': 1,
 'layer': None,
 'level': 6,
 'stroke_opacity': None,
 'dashes': None,
 'lineJoin': None,
 'lineCap': None,
 'color': None,
 'width': None}
```
The wheel is too big to attach, please download from here https://1drv.ms/u/s!AgGqVEdejC_KhOd_3KZCohsYznXmJg.

**Reply by akshaykarangale** [reply]
Thank you so much for this! You are the best :)

**Reply by JorjMcKie** [reply]
Thank you 😊, good luck with the test and **_please_** do provide feedback.

**Reply by akshaykarangale** [reply]
Hello, I tested the new version. It is much more accurate in extracting (and recreating) shapes than the previous versions. I have a question though - one of the PDFs that I was testing had a clipping on an image. I was able to pull the clip of the image and the image separately but could not find a way to link them together. Is it possible to identify which clip applies to which image? Thanks in advance :)

**Reply by JorjMcKie** [reply]
@akshaykarangale thanks for testing and your feedback - very valuable!

> I was able to pull the clip of the image and the image separately but could not find a way to link them together. Is it possible to identify which clip applies to which image?

Hm, thing is, that drawings (I learned the better name is "line art") are the only things contained in `page.get_lineart()`/`get_drawings()`.
The "data store" I am reading here however is the full MuPDF trace of the page, so contains everything - including image and pattern calls.
I **_might_** include these calls, so the `get_drawings(extended=True)` / `get_lineart()` would them too.

**_Or_** you "manually" match the image bbox with the clip scissors ... of course tedious and error prone:
* a clip scissor probably is smaller than the bboxes which it applies to.
* if there is a scissor hierarchy (the levels), then you would not have a way to find the right scissor, because the image bbox may intersect many in the scissor hierarchy, like so (indentation means level):
```
clip1
   clip2
      image
      clip3
         other stuff
```
"image" may intersect all three clips, but only the scissor intersection of clip1 and clip2 apply.

**Reply by JorjMcKie** [reply]
So to me it looks like I need to introduce at least one additional dictionary output, one for images plus maybe others for patterns or so.

**Reply by akshaykarangale** [reply]
Hello, thanks for the detailed reply. I think we will try to use the manual matching option until you introduce additional dictionaries. On a different note, is there any way I can get the "mutool trace" output using PyMuPDF library? Thanks :)

**Reply by JorjMcKie** [reply]
This is `mutool trace your.pdf > output.xml`. Not easy to read at first sight ...

**Reply by JorjMcKie** [reply]
You can also restrict to desired pages the same way as always: `mutool trace your.pdf 1,5 > page1-and-5.xml`.

**Comment by akshaykarangale** [comment]
Hi @JorjMcKie, I was wondering if you could give me a build for my production environment - Python 3.7.4 on Ubuntu 20.04.

Thanks in advance!
